### ISSUE TITLE
Need a way to let pictures "float" so they appear anywhere on the 2 pages near the key verse

### ISSUE THREAD

**Issue by davidc86** [sillsdev/ptx2pdf#56]
Currently the way pictures are handled in PTXprint is less than ideal, and getting a good looking output becomes very labor intensive on long documents (made up of several Bible books). It doesn't matter whether we specify t or b in the Pic List file, almost every picture causes huge gaps in the text. 
If program logic could peg the picture "somewhere near the key verse", so the picture ends up somewhere on the two facing pages where the key verse is, and the text just wraps across it, so that picture placement is independent of verses and paragraph marks, i.e. a paragraph can start above the picture and continue below it. I've only seen this kind of behavior in a very few cases (lucky placement it seems).
But if we hack the verse reference in the Pic List to point to a verse further down (e.g., the picture is talking about James 3:3, but I change the picture reference in Pic List to say JAS 3.7), then we can see that kind of "text flow" behavior in the PDFs . 
Unfortunately, we are having to do this on almost every picture.
(Why is this important to us? Where we work, there is no publications person, and we need to publish complete "Bible portion" books of the entire set of approved books for each language each time another book is approved. PTXprint comes very close already to making this doable by the local organization, Woohoo! But getting the pictures right creates a significant burden for them, e.g. the current Bible portions book for Selaru is over 700 A5 pages. So anything that can be done to get the output looking nice/something to be proud of, right out the gate, will be a huge help!)

**Comment by davidg-sil** [comment]
` It doesn't matter whether we specify t or b in the Pic List file, almost every picture causes huge gaps in the text.`
Really? This sounds like a bug - it should be doing what you're asking (sort of, it can't float to the 
previous page).... or are you saying that the verse you're specifying it in is one that would get pushed off the page if the picture was included? That would probably give you the behaviour you wanted.  If it's happening lots... just an idea, feel free to shout me down, but have you considered smaller pictures? e.g column width rather than page? or  a width of 0.7span? (ducks low and runs away)

**Comment by davidc86** [comment]
Ha! “ducks low and runs away” That was good.

Yes, I basically did a search and replace to make all Pic Lists specs say *.7 to try to minimize the problem. (BTW, a way to look at ALL the Pic Lists as a single document would have sped things up a bit.) And yes, this strategy did help significantly. Though some pictures look a bit tiny now, but these can be adjusted on an as needed basis.)

The main issue is, in a large book like Genesis or Acts or Luke, it is still very common for a picture at *.7 to still do exactly what you suspected: “that the verse you're specifying it in is one that would get pushed off the page if the picture was included” – yup. A LOT. And if we fix ONE picture in a large book so it doesn’t create a gap, then we have to recheck every following picture in that book, because now that text shift might have thrown a bunch of other pictures off too.

David

**Comment by mhosken** [comment]
Having pictures float across page boundaries is fraught with complexities. How far away should the figure be allowed to drift from its anchor? Should it allowed to be across a column and over a page? When should a figure float?

While it is desirable to have as many controls as possible to help with good page breaking, it is too much to ask the TeX macros to do more than they are doing already. The output routines are way too complicated as they stand, already.

This is not going to get fixed, at least with the limited understanding of writing this stuff that I have.

**Comment by davidc86** [comment]
I suspected we were pushing the limits on TeX already.

So, that is good to know.

It would also be good for something like that to be put in the FAQ, “Sorry but with the current abilities of the TeX engine this is as automated as we can get the picture placement to be. Finer tweaking will need to be done by the user in the book’s Pic List file. Here are some suggestions on how to get pictures to move some …”

David

**Comment by davidg-sil** [comment]
To add to Martin's comment, there is a TeX based system that does flow
pictures from page to page as is being asked for, and it has been around
for 30 years,  BUT: it was made and refined by some of the world"s  finest
TeXperts,  and it frequently breaks,  pushing pictures to the end of the
document. It works ok if there is never more than one picture per 3 pages
or so, but once picture one has been delayed, you get knock on effects. You
basically need to use the same approach, pushing picture definitions
earlier. The piclist approach is much nicer for this.

Typing without prior thought... I wonder how hard reliable DETECTION of the
issue might be, at picture insertion. Any thoughts, Martin?

**Comment by davidc86** [comment]
Though I find the picture placement to be "fiddly", it is doable, one picture at a time. I can successfully place pictures in reasonable relations to the "target" verse by looking at the results and seeing what verse (ahead or behind it) might be a better "attachment" verse. Then rerunning the process to see the results. Yes, a bit tedious, but it is consistently REPEATABLE! which is huge!  So I just need to take the time to tweak the PicList to get it right, and then it is good to go, whether printing one book or combining all the books. This is good enough. I don't suggest further messing with picture placement at this time.
